spec.html

```diff
@@ -6419,135 +6419,31 @@ <h1>GetGlobalObject ( )</h1>
-      <h1>EnqueueJob ( _queueName_, _job_, _arguments_ )</h1>
-      <p>The EnqueueJob abstract operation requires three arguments: _queueName_, _job_, and _arguments_. It performs the following steps:</p>
-      <emu-alg>
-        1. Assert: Type(_queueName_) is String and its value is the name of a Job Queue recognized by this implementation.
-        1. Assert: _job_ is the name of a Job.
-        1. Assert: _arguments_ is a List that has the same number of elements as the number of parameters required by _job_.
-        1. Let _callerContext_ be the running execution context.
-        1. Let _callerRealm_ be _callerContext_'s Realm.
-        1. Let _callerScriptOrModule_ be _callerContext_'s ScriptOrModule.
-        1. Let _pending_ be PendingJob { [[Job]]: _job_, [[Arguments]]: _arguments_, [[Realm]]: _callerRealm_, [[ScriptOrModule]]: _callerScriptOrModule_, [[HostDefined]]: *undefined* }.
-        1. Perform any implementation or host environment defined processing of _pending_. This may include modifying the [[HostDefined]] field or any other field of _pending_.
-        1. Add _pending_ at the back of the Job Queue named by _queueName_.
-        1. Return NormalCompletion(~empty~).
-      </emu-alg>
+    <p>Jobs are scheduled for execution by ECMAScript host environments with abstract operations such as HostEnqueuePromiseJob. Such operations accept two parameters: _job_, an abstract operation, and _arguments_, a list of ECMAScript values. These abstract operations schedule _job_ to be performed with _arguments_ at some future time. Their implementations must conform to the following requirements:</p>
+
+    <ul>
+      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
+        <ol>
+          <li>Push an execution context onto the execution context stack.</li>
+          <li>Perform any implementation-defined preparation steps.</li>
+          <li>Perform the abstract operation given by _job_, passing the arguments _arguments_.</li>
+          <li>Perform any implementation-defined cleanup steps.</li>
+          <li>Pop the previously-pushed execution context from the execution context stack.</li>
+        </ol>
+      </li>
+      <li>Only one Job may be actively undergoing evaluation at any point in time.</li>
+      <li>Once evaluation of a Job starts, it must run to completion before evaluation of any other Job starts.</li>
+    </ul>
+
+    <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
+      <h1>HostEnqueuePromiseJob ( _job_, _arguments_ )</h1>
+
+      <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the abstract operation indicated by _job_ to be performed, with the arguments _arguments_, at some future time. The Jobs used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
```

Comment:
are there no consistent steps needed here? I’d expect, at the least, an assertion about the arguments passed to the abstract operation.

Reply:
This text follows the pattern of other host hooks in not having algorithm steps--the algorithm is provided by the host. The argument types are documented in this paragraph.

Reply:
This paragraph does not define the types - there is zero text that explicitly says that `jobs` is a Job, and nothing that indicates that `arguments` is a List or an Array (i assume).

Reply:
The type of _job_ is given ("the abstract operation indicated by _job_"). I suppose you could state "with the arguments given by _arguments_ (a List)" if it's not clear enough from context.

Reply:
The argument name doesn’t guarantee what type it is.

Reply:
No, the type is instead guaranteed by the rest of the PR which gives a normative processing model that passes Lists.

Reply:
that's also true of a lot of other algorithms where we still use `Assert` (which I can attest really helps when implementing)

Reply:
@ljharb: The phrase `the abstract operation indicated by _job_` means that the parameter `_job_` is (bound to) an abstract operation. Its type *is* 'abstract operation'. (In the world of this PR, there isn't really a distinct Job type.)

As for `_arguments_`, you have to observe that `HostEnqueuePromiseJob` is a job-scheduling operation as described in the parent clause, which gives the parameter types of such operations more precisely.

Reply:
Thanks for the clarification - this seems like a really obscure way to specify things. For example, why not make a Job a record with an abstract operation in a slot, so it’s a reified “thing” with its own type? (not suggesting doing that; it’s just that this approach seems very vague and hand-wavey and hard to understand without the context of how the HTML spec works, which shouldn’t be required in order to understand this spec)

Reply:
I don't think wrapper types is a good strategy for increasing clarity; it just creates an extra layer of abstraction that makes it _less_ clear what is being operated on.

Reply:
I'd also point out that the job setup is inherited from the current spec, not something particularly tied to HTML. I agree that the current spec is pretty clear and close to how programming languages operate. E.g., `process.nextTick(job)` in Node.js, and similarly function pointers in C/C++.

Reply:
@ljharb:
> why not make a Job a record with an abstract operation in a slot, so it’s a reified “thing” with its own type?

In the world of this PR, you can think of that record existing in the host environment, but it's not the spec's job to specify the host environment. (At least, not to that level of detail.)

I agree with @domenic that introducing such a record to this PR (e.g., reinstating PendingJob) would not increase clarity.

> this approach seems very vague and hand-wavey

I think some of that is inevitable when the spec talks about the host environment and its interaction with the ES implementation. (On the other hand, I do think this PR could be clearer.)